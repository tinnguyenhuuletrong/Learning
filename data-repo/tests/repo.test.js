const { promisify } = require("util");
const { EventEmitter } = require("events");
const redis = require("redis");

/** @type {import('../@types/index')} */
const DataRepo = require("../src");
const {
  start,
  stop,
  createDataRepo,
  createDataRepoRemote,
  createRpcClient,
} = DataRepo;

const MemoryStore = require("../src/storages/MemoryStore");
const waitMs = (ms) =>
  new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve();
    }, ms);
  });

const URI = "redis://127.0.0.1:6379";

/** @type {import('redis').RedisClient} */
let redisClient;

const DummyDataSource = {
  findAsync: async () => {},
  on: () => {},
  off: () => {},
};

async function waitForConnect(client) {
  return new Promise((resolve, reject) => {
    client.once("connect", () => resolve(client));
    client.once("error", () => reject());
  });
}

beforeAll(async () => {
  redisClient = redis.createClient(URI);
  await waitForConnect(redisClient);
  await start(redisClient);
});

afterAll(async () => {
  redisClient.end(true);
  await stop();
});

test("DataRepo - rpc test", async () => {
  const ins = await createDataRepo({
    topic: "DomainA",
    dataSource: DummyDataSource,
  });
  await ins.start();

  const topic = ins.rpcServer.getRpcTopic();
  const client = await createRpcClient();

  const jobs = new Array(3).fill(1).map((k) =>
    client.request({
      ...topic,
      msg: {
        cmd: "ping",
      },
    })
  );

  expect(await Promise.all(jobs)).toMatchSnapshot();

  await ins.stop();
});

test("DataRepo - query data sources", async () => {
  const TestDataSource = {
    ...DummyDataSource,
    findAsync: async (query) => {
      return [
        {
          id: 0,
          name: "t1",
        },
        {
          id: 1,
          name: "t2",
        },
      ];
    },
  };

  const ins = await createDataRepo({
    topic: "DomainA",
    dataSource: TestDataSource,
  });
  await ins.start();

  const topic = ins.rpcServer.getRpcTopic();
  const client = await createRpcClient();

  const res = await client.request({
    ...topic,
    msg: {
      cmd: "find",
      query: {},
    },
  });

  expect(res).toMatchSnapshot();

  await ins.stop();
});

test("DataRepo - event", async () => {
  const TestDataSource = new EventEmitter();
  TestDataSource.findAsync = async () => [];

  const ins = await createDataRepo({
    topic: "DomainA",
    dataSource: TestDataSource,
  });
  await ins.start();

  const mockFunc = jest.fn();
  ins.stream.addToStream = mockFunc;

  TestDataSource.emit("created", {
    id: 2,
    name: "new",
  });

  TestDataSource.emit("updated", {
    id: 2,
    name: "new_1",
  });

  TestDataSource.emit("deleted", {
    id: 2,
  });

  expect(mockFunc).toMatchSnapshot();

  await ins.stop();
});

test("DataRepo - remote", async () => {
  const TestDataSource = new EventEmitter();
  TestDataSource.findAsync = async (query) => {
    return [
      {
        _id: 0,
        name: "t1",
      },
      {
        _id: 1,
        name: "t2",
      },
    ];
  };

  const topic = "DomainA";
  const memStore = new MemoryStore();
  const repoIns = await createDataRepo({
    topic,
    dataSource: TestDataSource,
  });
  await repoIns.start();

  const remoteIns = await createDataRepoRemote({
    topic,
    dataStore: memStore,
  });
  await remoteIns.start();

  console.log(memStore);
  expect(memStore.toObjects()).toMatchSnapshot({
    head: expect.any(String),
  });

  // Emit some change
  TestDataSource.emit("created", {
    _id: 2,
    name: "new",
  });
  TestDataSource.emit("updated", {
    _id: 1,
    name: "updated name",
  });
  TestDataSource.emit("deleted", {
    _id: 0,
  });

  await waitMs(1000);

  // Should sync
  console.log(memStore);
  expect(memStore.toObjects()).toMatchSnapshot({
    head: expect.any(String),
  });

  await repoIns.stop();
  await remoteIns.stop();
}, 10000);
