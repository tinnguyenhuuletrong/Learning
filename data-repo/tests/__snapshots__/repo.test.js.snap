// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRepo - event 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "evt": "created",
        "payload": Object {
          "id": 2,
          "name": "new",
        },
      },
    ],
    Array [
      Object {
        "evt": "updated",
        "payload": Object {
          "id": 2,
          "name": "new_1",
        },
      },
    ],
    Array [
      Object {
        "evt": "deleted",
        "payload": Object {
          "id": 2,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`DataRepo - query data sources 1`] = `
Object {
  "data": Array [
    Object {
      "id": 0,
      "name": "t1",
    },
    Object {
      "id": 1,
      "name": "t2",
    },
  ],
  "status": "success",
}
`;

exports[`DataRepo - remote 1`] = `
Object {
  "head": Any<String>,
  "items": Object {
    "0": Object {
      "_id": 0,
      "name": "t1",
    },
    "1": Object {
      "_id": 1,
      "name": "t2",
    },
  },
}
`;

exports[`DataRepo - remote 2`] = `
Object {
  "head": Any<String>,
  "items": Object {
    "1": Object {
      "_id": 1,
      "name": "updated name",
    },
    "2": Object {
      "_id": 2,
      "name": "new",
    },
  },
}
`;

exports[`DataRepo - rpc test 1`] = `
Array [
  Object {
    "data": "pong",
    "status": "success",
  },
  Object {
    "data": "pong",
    "status": "success",
  },
  Object {
    "data": "pong",
    "status": "success",
  },
]
`;
